src/main/java/com/jnape/palatable/lambda/io/MVar.java

```diff
@@ -0,0 +1,162 @@
+package com.jnape.palatable.lambda.io;
+
+import com.jnape.palatable.lambda.adt.Maybe;
+import com.jnape.palatable.lambda.functions.Fn1;
+
+import java.util.Collections;
+import java.util.concurrent.ArrayBlockingQueue;
+
+import static com.jnape.palatable.lambda.adt.Maybe.maybe;
+import static com.jnape.palatable.lambda.adt.Try.trying;
+import static com.jnape.palatable.lambda.functions.Fn0.fn0;
+import static com.jnape.palatable.lambda.functions.builtin.fn1.Constantly.constantly;
+import static com.jnape.palatable.lambda.io.IO.io;
+import static com.jnape.palatable.lambda.io.IO.throwing;
+
+/**
+ * A mutable single element memory cell with blocking operations {@link MVar#take()} and {@link MVar#put(A)};
+ *
+ * @param <A> The type of object the MVar contains
+ */
+public final class MVar<A> {
+
+    private final ArrayBlockingQueue<A> queue;
+
+    private MVar() {
+        queue = new ArrayBlockingQueue<>(1, true);
+    }
+
+    private MVar(A a) {
+        queue = new ArrayBlockingQueue<>(1, true, Collections.singletonList(a));
+    }
+
+    /**
+     * Construct a new empty {@link MVar}
+     *
+     * @param <A> The {@link MVar} parameter type
+     * @return An {@link IO} that produces the empty {@link MVar}
+     */
+    public static <A> IO<MVar<A>> newMVar() {
+        return io(fn0(MVar::new));
+    }
+
+    /**
+     * Construct a new occupied {@link MVar}
+     *
+     * @param a   The initial value
+     * @param <A> The {@link MVar} parameter type
+     * @return An {@link IO} that produces the occupied {@link MVar}
+     */
+    public static <A> IO<MVar<A>> newMVar(A a) {
+        return io(() -> new MVar<>(a));
+    }
+
+    /**
+     * Blocking retrieval of the value inside. If the {@link MVar} is currently empty the returned {@link IO} will block until it becomes occupied.
+     *
+     * @return An {@link IO} which will block and return a value from inside the {@link MVar}
+     */
+    public IO<A> take() {
+        return io(queue::take);
+    }
+
+    /**
+     * Blocking placement of the value inside. If the {@link MVar} is currently occupied the returned {@link IO} will block until it becomes empty.
+     *
+     * @return An {@link IO} which will block and return the {@link MVar} after mutation.
+     */
+    public IO<MVar<A>> put(A a) {
+        return io(() -> {
+            queue.put(a);
+            return this;
+        });
+    }
+
+    /**
+     * Read the contents inside without emptying.
+     * <p>
+     * Note that this operation is not atomic and could block on put if there are multiple producers.
```

Comment:
I can't find a way to make this atomic without either not using `ArrayBlockingQueue` or busy waiting on `Queue::peek`

Reply:
I'll merge it into it's own branch and add an internal atomic operational interface for these types of things: I'd like zero non-atomic compound operations on lambda's `MVar`.

Reply:
I have a sneaking suspicion that it will be simpler in the long run to dump `ArrayBlockingQueue` in favor of `AtomicReference` (or just a `volatile` field and direct `Unsafe` access), but I'll cross that bridge when/if I come to it. All in all a solid contribution, thank you.